### ISSUE TITLE
Disentangling initialization of optimizer from the parameters of the model to train

### ISSUE THREAD

**Issue by amine-ammor** [pytorch/pytorch#97603]
### 🚀 The feature, motivation and pitch

Is it possible and compatible with the architecture of pytorch to disentangle the initialization of on optimizer from the parameters of the model to train?

In pytorch, to initialize an optimizer, we need to define the 'params' argument with the optimizer's parameters.

For example, to define a SGD with a fixed learning rate and momentum over all the parameters, we need to execute a command, that would look as follows : 

`optimizer = torch.optim.SGD(params=model.parameters(), lr=0.00.1, momentum=0.9)`
which implies to define the model before defining the optimizer.

It would be interesting to be able to set the 'params' argument to None at initialization, before defining it afterwards.
For example using the "add_param_group" or a new method "link" , that would apply the defaults parameters over all the given parameters,
In this case the code above, would be replaced with:
`optimizer = torch.optim.SGD(None, lr=0.00.1, momentum=0.9)`
`optimizer.link(model.parameters())`

### Alternatives

For the moment the solution I thought of is the following,but it is a bit of a hack:

`empty_tensor = [torch.empty(0)]`
`optimizer = torch.optim.SGD(empty_tensor,lr=0.00.1, momentum=0.9)`
`optimizer.param_groups[0]["params"] = model.parameters()`

### Additional context

One use case of this feature, would to be define an optimizer, test it on a first model, then reuse-it again with the same parameters on another model (without having to redefine the optimizer or use the method I used above),
 for example using the following code:

`optimizer = torch.optim.SGD(None, lr=0.00.1, momentum=0.9)`
`optimizer.add_params(model1.parameters())`
`optimizer.unlink()`
`optimizer.add_params(model2.parameters())`
where unlink would reset the parameters to None.

One example of the use case of this feature, would be to test the optimizer with a first configuration (model or dataset of small size or toy model, to experiment with the optimizer ), then once validated to be reused again on a second configuration scaled up.
Another example, would be a pipeline for training, that would take as inputs , an optimizer and a model, that would link them inside ,which would allow to forward the responsability of linking them to the pipeline.

cc @vincentqb @jbschlosser @albanD @janeyx99

**Comment by amine-ammor** [comment]
In fact, there is a cleaner alternative using standard library:
```
from functools import partial
unlinked_optimizer = partial(torch.optim.SGD,lr=0.001)

optimizer1 = unlinked_optimizer(model.parameters())
optimizer2 = torch.optim.SGD(model.parameters(),lr=0.001)
assert optimizer1.state_dict() == optimizer2.state_dict()
```

**Comment by janeyx99** [comment]
It looks like the `partial` function already addresses the use case here of being able to pass along an optimizer configuration without the parameters. Is it also important that the internal state (e.g. the momentum buffers) to be preserved and re-used? In the use cases above, it seems as if the state sharing is not desired, but I may be missing something.

By the way, the way I've been doing something similar with my benchmarks is by storing a constructor and defaults config dictionary like:

```
from torch.optim import Adam, AdamW

defaults = {'fused': True, 'capturable': True, 'amsgrad': True}
Optim = Adam  # or if you're trying to test out different optims, this code go in a loop and the Optim and defaults can switch

optimizer1 = Optim(model1.parameters(), **defaults)
optimizer2 = Optime(model2.parameters(), **defaults)
```

This wouldn't require the partial function, but the partial function seems pretty clean. This also does NOT reuse the same optimizer instance.

**Comment by amine-ammor** [comment]
For the use-case specified above, state-sharing is not desired, but the link method could possess an additional argument that specify if the optimizer state should be preserved or not (I am not aware of use-cases, where we would change the model and dataset and keep the same internal state (momentum) unless maybe the change on the model and the dataset induces a continuous change of the loss ).

In fact, the aim of the requested feature is to offer more flexibility (without having to write additional code to share the same optimizer), it would generalize the actual interface, without breaking it behavior, but it would come with an obvious cost : for a more generic interface, we need to add more exception handlers (for example "step" and "zero_grad method" would need to verify if the optimizer if linked and raise a new exception if it is not the case  ) and  of course more tests.

**Comment by janeyx99** [comment]
Ah, in this case, since state is not shared and given the additional cost, we'd recommend using `partial` to achieve this purpose. I'll close the issue for now but please feel free to reopen if you find cases for which this isn't sufficient!
